### ISSUE TITLE
a3c_gcn_seq2seq's net seems does not work as paper describe

### ISSUE THREAD

**Issue by xuxi2569471573** [GeminiLight/virne#6]
hello, I find a3c_gcn_seq2seq's net seems does not work as paper describe.I found that placement decisions are directly determined by  logits = self.mlp(p_node_embeddings)，and context is not used.is this right? Could you please explain why or how to modify it？Thanks in advance!!! 
![image](https://user-images.githubusercontent.com/71211938/233953284-663de906-4662-4c48-ab59-720f003e1781.png)

**Comment by GeminiLight** [comment]
Before the final logits are generated (`logits = self.mlp(p_bode_embeddings)`), the p_node_embedding will be combined with the `hidden_state` of the current timestep. The `hidden_state` is a dynamic embedding for placement decisions as it is updated at each timestep by the attention `att` and GRU `gru` modules. 

By placing the code responsible for generating the logits after the hidden state update, we can see that the implementation matches the description provided in the paper.

I hope that providing this explanation will clear up the confusion you have. And If you have any questions, feel free to ask.

**Comment by xuxi2569471573** [comment]
I'm so sorry I didn't find the hidden_ State is updated by attention att .The hidden_ Stat was passed into attention att as a parameter and don't have changes like below. Could you please explain it?
![image](https://user-images.githubusercontent.com/71211938/233987965-7bedd827-dbea-4281-b263-17ca1fcae69c.png)

**Comment by GeminiLight** [comment]
Thank you for bringing to my attention the issue with the current implementation. I understand that both the `hidden_state` and `context` contain information regarding the placement decision sequence. However, by combining the `p_node_embeddings` with the `context` instead of the `hidden_state`, we can utilize the attention mechanisms to capture more features. I appreciate your feedback and will address this bug in the upcoming version. Thank you once again.

Code:
```python
    def forward(self, obs):
        batch_p_net = obs['p_net']
        hidden_state = obs['hidden_state']
        p_node_embeddings = self.gcn(batch_p_net)
        p_node_embeddings = p_node_embeddings.reshape(batch_p_net.num_graphs, -1, p_node_embeddings.shape[-1])
        p_node_id = obs['p_node_id']
        hidden_state = hidden_state.permute(1, 0, 2)
        encoder_outputs = obs['encoder_outputs']
        mask = obs['mask']
        p_node_emb = self.emb(p_node_id).unsqueeze(0)
        context, attention = self.att(hidden_state, encoder_outputs, mask)
        context = context.unsqueeze(0)
        outputs, hidden_state = self.gru(p_node_emb, hidden_state)
        p_node_embeddings = p_node_embeddings + context # or p_node_embeddings + context + hidden_state
        logits = self.mlp(p_node_embeddings)
        return logits, outputs, hidden_state
```

**Comment by xuxi2569471573** [comment]
Thank you for your reply. I would like to ask why not concatenate them like paper describe and instead add them. Is this more effective？

**Comment by GeminiLight** [comment]
Concatenation `concat` and addition `add` are commonly used fusion modules in deep learning. They are theoretically almost equivalent, with only slight differences in their results. However, adding two tensors avoids the need for dimension expansion, making it a more efficient option in most cases.
